### PR TITLE
Conditional valves (Formerly Pressure)

### LINKED ISSUE #33709: [Feature Request] Temperature valves? Conditional valves?
[featurerequest][discussion]

I'll start by relaying this feature request I saw on the forums:
>"_I would love some temperature valves similar to the pressure ones that would allow it to open or close once the connected pipe exceeds or drops below a certain temperature._" - falcon2346

For starters I'd like to gauge how much interest there is in temperature valves and potentially other kinds of conditional valves. I'd like to keep the number of conditional valve types to a minimum (something like a gas type valve can be done already without too much fuss using a pressure valve and a gas filter) but still, suggestions welcome.

That said, my main concern is implementation: Should there be entirely new conditional valve devices bloating the RPD's list, or should the pressure valve become an universal "conditional valve" instead, where you can set what parameter to check against through the valve's menu?
The later would mean figuring out ways to visually determine what mode (pressure, temperature, whatever else) a conditional valve has been set to, probably through relatively minor sprite differences, while the former option would allow for visually distinct sprites at the cost of bloating the RPD's list. 
...Which might not be that much of an issue if it's just the temperature valves for now, I suppose?

I don't know. Any ideas on this whole thing are welcome.

-- 

New points of discussion raised so far:
* In relation to RPD bloat concerns, the RPD's list could use some TLC, such as  tidying up the categories or creating dropdows.
* Possible additions to conditional valve behaviour in general. Suggestions so far: Ability to compare the control against the input pipe, rather than just the constant.

### PR DESCRIPTION
**Pull request by jellyveggie2**
## What this does
Adds some extra stuff to pressure valves, which have been renamed to conditional valves:
* You can now compare temperature instead of pressure.
* You can have them read the threshold from any of the input pipes.
* You can set any of the three pipes as the control pipe, making the side-pipe optional and allowing you to "inline" valves.

valveui2

There's been some sprite changes to reflect this:

valves

* A blue or orange light indicates the valve's comparison mode (pressure or temperature respectively)
* A light or a switch on the input pipes pointing to the side-pipe means that pipe is being read, either as control or as threshold (Couldn't really figure out a way to make it clear which).

Closes #33709

(Could have sworn there was a [content] tag, but I can't find it on the wiki)

## Why it's good

Extra functionality to a new-ish toy, for supermatter and plasma cooling autism

## Changelog

:cl:
 * rscadd: Pressure valves can now also compare temperature, and have been renamed to conditional valves.
 * tweak: Conditional valves can now also compare the control pipe against the input pipes' gases, rather than just a fixed value.
 * tweak: Conditional valves can now also forego the side input entirely, using either of the two other pipes as control. This might help make some designs neater and more compact.
 * rscdel: Conditional valves can no longer be controlled via AAC, beyond simple on/off commands.

### CODE REVIEW

**Comment by Inorien** [comment]
Why does this need to replace the pressure valve? Though I'll admit I have never used a pressure valve so if they were really that dumb and useless then this is cool, no idea what the fuck is going on in the sprite tho that shit is complicated and huge

The naming of up/down/left whatever constantly comes up, at the moment the gas mixer interface refers to things by north/east/south/west so may be better to keep consistency with that rather than introduce another scheme

Also your changelog is referring to itself (?)

**Comment by pedr69** [comment]
it's fucking happening
non positronic AI circuits when

**Comment by Eneocho** [comment]
Atmospherics is a subtle joke. The pipe is called "Conditional" Valves, where conditional ends with "-tional", thus not rhyming with the name of the department, Atmospherics. The title says that the pipe was "Formerly Pressure", implying that the word beginning with "C" would've ended with "-sure". So, before this PR was opened, it would've been called "Condisure Valves".

**Comment by jellyveggie2** [comment]
@Inorien 
Well... At first I tried to do the RPD clean-up/QoL mentioned in #33709 but failed, and decided I would focus on the valve stuff instead. So to compensate, out of the two options discussed in that issue ("make two separate valves for temperature and pressure at the cost of RPD clutter" vs "make a single valve that can do both temperature and pressure") I choose the "single valve" option.
None of the original functionality is lost, in fact it already defaults to what used to be the original functionality (pressure mode, center control, constant threshold), so it's not that much of a replacement anyway other than the new name, since it's not just pressure anymore. 

You've got a valid point with the North/South thing and in fact that's what I first tried to do, with names updating on the UI depending on valve position, but then came updating the AAC's menu and it all started breaking apart because I would no longer have access to the valve's rotation and update the North/South/whatever inputs' direction accordingly (moreover, an instruction on the AAC can refer to multiple valves, each with their rotation).
I was going to have to come up with new terminology no matter what, and figured keeping consistent would be the lesser of to evils (doing so also simplified UI code, and I was starting to look for ways to be lazy and finish this thing already at this point). 
I will admit my naming sense sucks and I am open to alternatives.

And yeah, the new sprites do look huge now. That's what I get trying to add extra pipes in an attempt at showing where gas is being redirected. The height of my spriting skill stops at gluing pieces of other sprites together, so if an actual spriter comes in with something better I'll take it.

> Also your changelog is referring to itself (?)

...Is that really that bad? The lines are meant to be read together, they're not going to get mixed up with other changes, and I'd rather not keep repeating myself about pressure valves now being called conditional valves.

**Comment by pedr69** [comment]
>but then came updating the AAC's menu and it all started breaking apart

isn't the point of these things not needing an AAC script?
surely you could get a functional conditional valve(albeit much bigger) with a script and some pumps, meters and valves

**Comment by Eneocho** [comment]
You could use these to make logic gates.
Connect the gates to the AAC, and the AAC to signalers.
Connect the signalers to light tile remotes.
You now have a screen.
Code tetris on space atmos.

**Comment by jellyveggie2** [comment]
>isn't the point of these things not needing an AAC script?

I made the mistake of making pressure valves compatible with the AAC thinking "why not", and I now have to live with the consequences.

**Comment by Inorien** [comment]
>but then came updating the AAC's menu and it all started breaking apart
this sounds like suffering
yeah as others have said not sure it needs AAC, sure you could get bonkers complexity with it but may not be worth

**Comment by Eneocho** [comment]
AAC valves please for the ultrahyperautismo.

**Comment by jellyveggie2** [comment]
Well, that's a 2 to 1 on dropping AAC support so far. 
I'll look into getting the North/South stuff back this evening unless the emojicracy votes otherwise on this comment:

Drop AAC support for conditional valves, yay or nay?

**Comment by jellyveggie2** [comment]
Wait I miscounted pedr69's as Eneocho changing his vote, the ratio's still 2 to 1.

Not that it changes the result I guess:
Removed AAC support for the valve (except the main enable/disable control, that one's simple enough it can stay) in favor of a better UI, and updated the PR accordingly. I'm still left with a need to refer to the pipes in a direction agnostic way, even if just for the changelog, so "side pipe" and "input pipes" will have to do for now.

**Comment by Eneocho** [comment]
AACbros we got too cocky...

File: code/ATMOSPHERICS/components/trinary_devices/pressure_valve.dm [hunk 1]
```diff
@@ -76,40 +136,74 @@
...
 	. = ..()
 	if (!enabled)
 		close()
 		return
 
-	if (open_on_above_threshold && (air2.pressure > pressure_threshold) || !open_on_above_threshold && (air2.pressure < pressure_threshold))
+	var/input = 0
+	var/threshold = 0
+	switch(control_input)
+		if (CVALVE_INPUT_LEFT) input = get_valve_mode_gas_value(pipe_flags & IS_MIRROR ? air3 : air1)
```
Comment (ShiftyRail) [inline code/ATMOSPHERICS/components/trinary_devices/pressure_valve.dm]: Formatting, we don't do one-line `if`s, this is bad for reading; especially with the weird ternary operator after

File: code/ATMOSPHERICS/components/trinary_devices/pressure_valve.dm [hunk 2]
```diff
@@ -76,40 +136,74 @@
...
+		if (CVALVE_INPUT_CENTER) input = get_valve_mode_gas_value(air2)
+		if (CVALVE_INPUT_RIGHT) input = get_valve_mode_gas_value(pipe_flags & IS_MIRROR ? air1 : air3)
+		else
+			control_input = CVALVE_INPUT_CENTER
+			input = get_valve_mode_gas_value(air2)
+
+	switch(threshold_source)
+		if (CVALVE_THRESHOLD_LEFT) threshold = get_valve_mode_gas_value(pipe_flags & IS_MIRROR ? air3 : air1)
```
Comment (ShiftyRail) [inline code/ATMOSPHERICS/components/trinary_devices/pressure_valve.dm]: Ditto

File: code/ATMOSPHERICS/components/trinary_devices/pressure_valve.dm [hunk 3]
```diff
@@ -76,40 +136,74 @@
...
+			threshold = constant_threshold
+
+	if (open_on_above_threshold && (input > threshold) || !open_on_above_threshold && (input < threshold))
 		open()
 	else
 		close()
 
+/obj/machinery/atmospherics/trinary/pressure_valve/proc/get_valve_mode_gas_value(datum/gas_mixture/air)
+	switch(mode)
+		if (CVALVE_MODE_PRESSURE) return air.pressure
```
Comment (ShiftyRail) [inline code/ATMOSPHERICS/components/trinary_devices/pressure_valve.dm]: Ditto

**Review by ShiftyRail — CHANGES_REQUESTED** [review]
Minor formatting changes as far as I can see

File: code/modules/atmos_automation/implementation/digital_pvalves.dm [hunk 4]
```diff
@@ -1,96 +1,43 @@
...
+				cvalves |= V.id_tag
+		if(cvalves.len == 0)
+			to_chat(usr, "<span class='warning'>Unable to find any conditional valves on this frequency.</span>")
 			return
-		pvalve = input("Select a valve:", "Sensor Data", pvalve) as null | anything in pvalves
-		parent.updateUsrDialog()
-		return 1
-
-// --- Pressure Valve Mode: Threshold and Above/Below ---
-/datum/automation/set_pressure_valve_mode
```
Comment (ShiftyRail) [inline code/modules/atmos_automation/implementation/digital_pvalves.dm:46]: Not familiar with those, where did they go? Does it result in feature loss for the AAC?

Comment (jellyveggie2) [reply code/modules/atmos_automation/implementation/digital_pvalves.dm:46]: That is right.
`/datum/automation/` are the commands that show up on the AAC. and `/datum/automation/set_pressure_valve_mode` would be the command I added when I first coded pressure valves to control their settings through the AAC. So yes, deleting that command would result in some feature loss. 
...I should note that in the changelog.

As to what led to this: #33709 brought up a request to 
>allow comparison between the control pipe (the one that is used for the conditional check) and input pipe (the one that is opened to the other side)

And that meant being able to select which input to compare the control pipe against (as there are two inputs that), which in turn meant naming each pipe. 
Applying the North/South/East/West naming scheme used in other atmos components such as gas mixers onto the valve's menu was easy enough... But it did not mesh well when it came to the AAC: as far as I can tell there's no way for me to read the controlled valve's `dir`, nor would it make sense as I must take into account the possibility that the AAC may be controlling multiple valves (each with their own `dir`).

So it's either:
* Adopt a new, `dir`-agnostic naming scheme specifically for conditional valves (my first choice), or
* Drop AAC support entirely (the choice that won out after some brief discussion in this PR)

You may want to look at [3bbba54](https://github.com/vgstation-coders/vgstation13/pull/34418/commits/3bbba54854e09b1067e4b1adca4f023538b14f0b) for a comparison between the two. I am open to alternative solutions as well.

Comment (jellyveggie2) [reply code/modules/atmos_automation/implementation/digital_pvalves.dm:46]: Forgot to mention, there _is_ a third option in that I could have the AAC and the valve follow their own naming schemes each, but I really don't like the lack of consistency there.

**Review by ShiftyRail — APPROVED** [review]

**Comment by ShiftyRail** [comment]
I /think/ AAC-independant valves are not too bad in and of themselves (we have manual valves for a reason), but maybe someone in the future will want to have their own way of tempering with them from a distance. We'll see.

**Pull request merged**